AP_BattMonitor: Add FuelLevel_Analog support

TunaLobster opened the pull request:
Adds a BattMonitor driver to use analog fuel sensors. Such as SAE fuel senders that output resistance or voltage. Tested on a CubeOrange with a potentiometer as the fuel level sensors on pin 103.

![FuelLevelAnalog](https://user-images.githubusercontent.com/5710309/174665273-74416797-09d5-45f3-94f8-1e93942ac59b.PNG)

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    float voltage = _analog_source->voltage_average();
+    uint32_t now_us = AP_HAL::micros();
```
```suggestion
    const uint32_t now_us = AP_HAL::micros();
```

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,62 @@
+        return;
+    }
+
+    float voltage = _analog_source->voltage_average();
```
```suggestion
    const float voltage = _analog_source->voltage_average_ratiometric();
```

Possibly should use ratiometric instead depending upon the sensor type.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,62 @@
+    /*
+      this driver assumes that CAPACITY is set to tank volume in millilitres.
+     */
+    float tank_level_ratio = voltage / 3.3;
```
Pretty sure we need a scalar ratio for these, as there may be external dividers in front of this that we can't control, and to correct for scaling errors from divider components.

WickedShell replied:
Also needs to cope with an input that works inside of the not full range (for example I've got a sensor on the way that has 0.25 - 4.75V output for the fuel level, which I can use an ADC to cut down to the right voltage range, but I need to be able to chop off the bottom bit of the range or it will never detect an empty tank correctly.

TunaLobster replied:
My other thought was to use read_average to skip the pin_scalar.

WickedShell replied:
You don't want to just deal with the internal scalers, there's also external scalars that we might have to add, and we won't be able to perfectly map to the 0 - 3.3V range, especially not if we want some error tolerance.

TunaLobster replied:
A couple new parameters for low/high voltage and linear interpolation? (My end goal with all of this is to get a CAN node to output FuelTankStatus, but need the ground work first.)

WickedShell replied:
That sounds perfect to me. That should solve the ratio thing I pointed out above with fewer parameters as well. It also makes it easy to calibrate because you can just fill it all the way, and drain it.

Maybe actually just stick the current ADC pin reading as the voltage because it makes easy debug?

WickedShell reviewed (changes requested):
I'm excited to see this, because it was pending on my to do list.

tridge reviewed (changes requested):
I think we'll need a low pass filter, maybe 0.2Hz cutoff?

WickedShell commented:
> I think we'll need a low pass filter, maybe 0.2Hz cutoff?

I'm pretty sure the sensor I'm looking at is internally filtered, but obviously not all sensors are, and for gas applications I don't think we will notice any problems from a filter that slow. (Maybe 0.5Hz, but that might be too fast?)

TunaLobster commented:
Built and tested on Cube Orange pin 103.

- Added a low pass filter at 0.3Hz. 0.2Hz seemed a little slow and 0.5Hz seemed a little fast for a typically slow speed signal. So I put it on the safe side of the middle.
- Used _volt_offset for empty voltage and _volt_multiplier for full voltage. Slightly combersome, but not additional parameter space is taken up. The AP_Float values were already there.
- Added the ADC value to the voltage field and the filtered value to the current_amps.
- Linear interpolation between low and high voltage.
- Added an option to enable the low pass filter.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,74 @@
+    // the empty voltage is _volt_offset
+    // the full voltage is _volt_multiplier
+    const float &voltage_used = (uint32_t(_params._options.get()) & uint32_t(AP_BattMonitor_Params::Options::FuelLevel_Analog_LPF))?filtered_voltage:voltage;
+    const float fuel_level_ratio = (voltage_used - _volt_offset) / (_volt_multiplier - _volt_offset);
```
Why are we subtracting the volt offset from the multiplier? That seems counter intutive to me, I was expecting this to be: `(voltage_used - _volt_offset) * _volt_multiplier`

TunaLobster replied:
I chose to go with the approach of setting both points and figuring out the slope rather than trying to fuss with setting a slope as a user. Both parameters can be set by reading the values from the GCS with empty and full tanks.

WickedShell replied:
It just seems odd to me that fine tuning your empty point will actually throw off any scaling correction you've done/know. It also just stands out as being done very different then the analog battery current monitor which does it as I outlined above, and as a user familiar with the analog battery I would have expected this to look more like that.

WickedShell replied:
Particularly to do it as you've implemented I think needs a lot more in terms of additions to the parameter documentation.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,74 @@
+    // the full voltage is _volt_multiplier
+    const float &voltage_used = (uint32_t(_params._options.get()) & uint32_t(AP_BattMonitor_Params::Options::FuelLevel_Analog_LPF))?filtered_voltage:voltage;
+    const float fuel_level_ratio = (voltage_used - _volt_offset) / (_volt_multiplier - _volt_offset);
+    const float fuel_level_used_ratio = 1.0 - fuel_level_ratio;
```
Do we need to clamp this at all? It seems pretty easy to get to an overfull/empty signal here. Which on the one hand seems bad to mask, on the other, it's clearly erroneous data...

TunaLobster replied:
I did think about that, but I would rather have a tank that get "overfull" and notice it during pre-flight checks rather than have that issue masked and have an issue in flight.

TunaLobster commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,77 @@
+
+    // get votlage from an ADC pin and filter it
+    const float voltage = _analog_source->voltage_average();
+    const float filtered_voltage = _voltage_filter.apply(voltage, float(dt_us) / 1000.0);
```
Should have remembered that one from the last PR that went through and changed a ton of those. Fixed.

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,74 @@
+
+    // get votlage from an ADC pin and filter it
+    const float voltage = _analog_source->voltage_average();
+    const float filtered_voltage = _voltage_filter.apply(voltage, float(dt_us) * 1e-3f);
```
needs to be 1.0e-6 to get it in seconds

tridge replied:
a param table for this backend would make sense, and add a filter frequency in Hz as an AP_Float

peterbarker commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -274,6 +275,9 @@ AP_BattMonitor::init()
             case Type::FuelLevel_PWM:
                 drivers[instance] = new AP_BattMonitor_FuelLevel_PWM(*this, state[instance], _params[instance]);
                 break;
+            case Type::FuelLevel_Analog:
```
#if AP_BATTMON_SMBUS_ENABLE

TunaLobster replied:
Macros created for each of the fuel features. All gated off from 1MB boards. Added the options to the build server too.

tridge commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,74 @@
+*/
+void AP_BattMonitor_FuelLevel_Analog::read()
+{
+    if (!_analog_source->set_pin(_curr_pin)) {
```
needs to check that analog_source exists, so it doesn't dereference nullptr

TunaLobster replied:
I think I fixed this. Looked around at other pieces of code that did this by `_analog_source ==nullptr`, correct?

TunaLobster commented:
I'll wait for the reviews before asking for the flapping CI tests to be re-run.

TunaLobster commented:
Changed the filter toggle to be slightly more helpful. Retested on a cube orange.

hendjoshsr71 commented on libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp:
```diff
@@ -0,0 +1,118 @@
+    // @Range: 0.01 10
+    // @Units: V
+    // @User: Advanced
+    AP_GROUPINFO("FL_VLT_MIN", 50, AP_BattMonitor_FuelLevel_Analog, _fuel_level_empty_voltage, 0.5),
```
Is there a backend using 40-49? Already? I didn't see it. 

We should probably add a comment table for this somewhere higher up?
We are near the end soon and will need to extend it. 

```suggestion
    AP_GROUPINFO("FL_VLT_MIN", 40, AP_BattMonitor_FuelLevel_Analog, _fuel_level_empty_voltage, 0.5),
```

hendjoshsr71 replied:
i looked and didn't find a 40-49 in the backends of master? 

By comment table somehwere that says

```cpp
// Paramter Indices for Parameters Loaded By Pointer
// Analog 1-9
// SMBUS 10-19
// INA2xx.cpp 25-26 (comments for this and below need to be fixed)
// UAVCAN 20-29
// Sum 30-39
// Fuel_Level_Analog 40-49
```

TunaLobster replied:
Fixed. Figured out where I got confused. The 40 range was used for the AP_SUBGROUPVARPTR instances.

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/skyviper-v2450/hwdef.dat:
```diff
@@ -73,8 +73,7 @@ define GPS_MOVING_BASELINE 0
 define HAL_ADSB_UAVIONIX_MAVLINK_ENABLED 0
-define AP_AIS_ENABLED 0
-define AP_BATTMON_FUEL_ENABLE 0
+define HAL_AIS_ENABLED 0
```
needs to be AP_AIS_ENABLED

TunaLobster replied:
Not sure when or why I might have done that. Fixed
